tools/xtce_generation/telemetry_generator.py

```diff
@@ -26,298 +64,280 @@ class TelemetryGenerator:
-        self.apid = apid
-        self.source_link = "http://www.omg.org/space/xtce"
-
-        if pkt is None:
-            # Read excel sheet
-            xls = pd.ExcelFile(path_to_excel_file)
-            self.pkt = xls.parse(packet_name)
-        else:
-            self.pkt = pkt
-
-    def create_telemetry_xml(self):
-        """
-        Create an XML representation of telemetry data based on input parameters.
-
-        Returns:
-        - root: The root element of the generated XML tree.
-        - parameter_type_set: The ParameterTypeSet element.
-        - parameter_set: The ParameterSet element.
-        - telemetry_metadata: The TelemetryMetaData element.
-        """
+    def __init__(self, path_to_excel_file):
+        # Read in all sheets from the excel file
+        self.sheets = pd.read_excel(path_to_excel_file, sheet_name=None)
+        # Set up the packet mapping from packetName to Apid
+        packet_sheet = self.sheets["Packets"]
+        self._packet_mapping = packet_sheet.set_index("packetName")["apId"].to_dict()
+
+        # Create the XML containers that will be populated later
+        self._setup_xml_containers()
+        # Add the CCSDS Header information to the containers
+        self._setup_ccsds_header()
+        # Create the sequence containers (also adding parameters within)
+        self._create_container_sets()
+
+    def _setup_xml_containers(self):
+        """Create an XML representation of telemetry data."""
         # Register the XML namespace
-        Et.register_namespace("xtce", self.source_link)
-
-        # Get the current date and time
-        current_date = datetime.now().strftime("%Y-%m")
+        source_link = "http://www.omg.org/space/xtce"
+        Et.register_namespace("xtce", source_link)
 
         # Create the root element and add namespaces
         root = Et.Element("xtce:SpaceSystem")
-        root.attrib["xmlns:xtce"] = f"{self.source_link}"
-        root.attrib["name"] = self.packet_name
-        # xmlns:xtce="http://www.omg.org/space/xtce"
+        self._root = root
+        root.attrib["xmlns:xtce"] = source_link
+        # Subsystem sheet name is used as the base name for this XTCE definition
+        subsystem = self.sheets["Subsystem"]
+        root.attrib["name"] = subsystem.loc[
+            subsystem["infoField"] == "subsystem", "infoValue"
+        ].values[0]
 
         # Create the Header element with attributes 'date', 'version', and 'author'
         # Versioning is used to keep track of changes to the XML file.
         header = Et.SubElement(root, "xtce:Header")
-        header.attrib["date"] = current_date
-        header.attrib["version"] = "1.0"
+        header.attrib["date"] = str(
+            subsystem.loc[
+                subsystem["infoField"] == "sheetReleaseDate", "infoValue"
+            ].values[0]
+        )
+        header.attrib["version"] = subsystem.loc[
+            subsystem["infoField"] == "sheetReleaseRev", "infoValue"
+        ].values[0]
```

Comment:
nice. with this, we get true date of when packet definition was updated last according to instrument team.